Docs: Optimized introduction of OIDC Bearer Token Authorization document (#27432)
quarkusio/quarkus: docs/src/main/asciidoc/security-openid-connect.adoc

include::./attributes.adoc[]
:toc:

You can use the Quarkus OpenID Connect (OIDC) extension to secure your JAX-RS applications using Bearer Token Authorization. 

> The Bearer Tokens are issued by OIDC and OAuth 2.0 compliant authorization servers, such as https://www.keycloak.org[Keycloak]. 

Reviewer: @hmanwani-rh 
Capitalization of "Tokens" here. I think this should be a lower case "t" in this instance as it is a more generic reference. Also, in the phrase "Bearer Token authorization", "authorization" is in lower case

Reviewer: @hmanwani-rh : The title doesn't appear to be changed here. Do you want to modify the title in this phase per our style guidelines?

Reviewer 2: @hmanwani-rh @sheilamjones If we are OK with `Authorization Code Flow` then indeed, I guess it makes sense to keep it consistent, `Bearer Token Authorization` is a concept, a single piece... I know we have in some places  `Bearer token authorization` - may be we should do a dedicated review of all these expressions in another PR...

Author: Thanks @sberyozkin

> Bearer Token authorization authorizes HTTP requests based on the existence and validity of a Bearer Token. 

Reviewer 2: ```suggestion
Bearer Token Authorization is the process of authorizing HTTP requests based on the existence and validity of a Bearer Token as well as whether an HTTP resource can be accessed.

Author: Done

> The Bearer Token provides information about determining the subject of the call and whether or not an HTTP resource is accessible. 

Reviewer 2: ```suggestion
The Bearer Token provides information about the subject of the call which is used to determine whether or not an HTTP resource can be accessed. 
```

Author: Done

> If you want to authenticate and authorize the users using OpenID Connect Authorization Code Flow, see xref:security-openid-connect-web-authentication.adoc[Using OpenID Connect to Protect Web Applications].

Reviewer 2: ```suggestion
If you need to authenticate and authorize the users using OpenID Connect Authorization Code Flow, see xref:security-openid-connect-web-authentication.adoc[Using OpenID Connect to Protect Web Applications].
```

Reviewer: @hmanwani-rh: Propose to say "by" using OpenID Connect Authorization Code Flow,...."

Author: Done

include::./attributes.adoc[]
:toc:

You can use the Quarkus OpenID Connect (OIDC) extension to secure your JAX-RS applications using Bearer Token Authorization. 
The Bearer Tokens are issued by OIDC and OAuth 2.0 compliant authorization servers, such as https://www.keycloak.org[Keycloak]. 

Bearer Token Authorization is the process of authorizing HTTP requests based on the existence and validity of a Bearer Token. 
The Bearer Token provides information about the subject of the call which is used to determine whether or not an HTTP resource can be accessed.

> The following figures outline the Bearer Token Authorization mechanism in Quarkus:

Reviewer 2: ```suggestion
The following diagrams outline the Bearer Token Authorization mechanism in Quarkus:
```

Reviewer 2: `diagram` seems a bit more technical, and the code flow doc also uses the `following diagram`

Author: Done

.Bearer Token Authorization mechanism in Quarkus

> image::security-bearer-token-authorization-mechanism-1.png[alt=Bearer Token Authorization, align=center]

Reviewer 2: ```suggestion
image::security-bearer-token-authorization-mechanism-1.png[alt=Bearer Token Authorization,width="60%", align=center]
```

Reviewer 2: Lets try Martin's suggestion with these images, see #27876

Author: Done

> In the previous figure:

Reviewer 2: ```suggestion
Figure 1. Bearer Token Authorization mechanism in Quarkus
```

Reviewer 2: This is how it is done at https://quarkus.io/version/main/guides/security-openid-connect-web-authentication (note it is in italics), so I think it has to be consistent; May be it should be `Diagram 1. ...`

Author: The `Figure 1....` is actually the caption for the image that we add to the diagram. 

How about changing it to `In the previous diagram`:

Reviewer 2: @hmanwani-rh Hi, I'd rather have it consistent with the code flow doc, I'm not quite sure `in the previous...` is a typical way, may be we should just remove `In the previous...` and have no `Figure 1` and just start enumerating what is going on in the diagram ?

Reviewer 2: @hmanwani-rh Thinking more about it, it does look better to me with this `Figure 1....` as we have 2 ones here and each one has a different description

> . First, the Quarkus service retrieves verification keys from the OpenID Connect provider.

Reviewer 2: ```suggestion
. 1. The Quarkus service retrieves verification keys from the OpenID Connect provider. It will use them to verify the bearer access tokens.
```

Author: Done, changed to following sentence:

The Quarkus service retrieves verification keys from the OpenID Connect provider. The verification keys are used to verify the bearer access token signatures.

> . The Quarkus user accesses the single-page application.

Reviewer 2: ```suggestion
2. The Quarkus user accesses the Single Page Application.
```

Author: Done

> . The single-page application uses Authorization Code Flow to authenticate the user and retrieve tokens from the OpenID Connect provider.

Reviewer 2: ```suggestion
3. The Single-page application uses Authorization Code Flow to authenticate the user and retrieve tokens from the OpenID Connect provider.
```

> . The single-page application uses the access token to retrieve the service data from the Quarkus service.

Reviewer 2: ```suggestion
4. The Single-page application uses the access token to retrieve the service data from the Quarkus service.
```

Author: Done

> . The Quarkus service verifies the bearer access token and returns data to the single-page application.

Reviewer 2: ```suggestion
5. The Quarkus service verifies the bearer access token signature using the verification keys, checks the token expiry date and other properties, allows the request to proceed and returns the service response data to the Single-page application.
```

Author: Done

> . The single-page application returns the same data to the Quarkus user.

Reviewer 2: ```suggestion
6. The Single-page application returns the same data to the Quarkus user.
```

Author: Done

.Bearer Token Authorization mechanism with Client in Quarkus

> image::security-bearer-token-authorization-mechanism-2.png[alt=Bearer Token Authorization, align=center]

Reviewer 2: ```suggestion
image::security-bearer-token-authorization-mechanism-2.png[alt=Bearer Token Authorization, width="60%", align=center]
```

Author: Done

> In the previous figure:

Reviewer 2: ```suggestion
Figure 2. Bearer Token Authorization mechanism in Quarkus with Java or command line client.
```

> . First, the Quarkus service retrieves verification keys from the OpenID Connect provider.

Reviewer 2: ```suggestion
1. The Quarkus service retrieves verification keys from the OpenID Connect provider. It will use them to verify the bearer access token signatures.
```

> . The Client uses provided credentials to retrieve the access token from the OpenID Connect provider.

Reviewer 2: ```suggestion
. The Client uses a client_credentials which requires a client id and secret or password grant which requires a client id and secret as well as a user name and password to retrieve the access token from the OpenID Connect provider.
```

Author: Changed to following sentence:

2. The Client uses `client_credentials` that requires client ID and secret or password grant, which also requires client ID, secret, user name, and password to retrieve the access token from the OpenID Connect provider.

> . The Client uses the access token to retrieve the service data from the Quarkus service.

Reviewer 2: ```suggestion
3. The Client uses the access token to retrieve the service data from the Quarkus service.
```

Author: Done

> . The Quarkus service verifies the bearer access token and returns data to the Client.

Reviewer 2: ```suggestion
4. The Quarkus service verifies the bearer access token signature using the verification keys, checks the token expiry date and other properties, allows the request to proceed if the token is valid and returns the service response to the Client.
```

Author: Done
